## Optimize the Model


In [ ]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

#  Import and read the charity_data.csv.
import pandas as pd 
application_df = pd.read_csv("charity_data.csv")
application_df.head()

EIN                                      NAME  ... ASK_AMT IS_SUCCESSFUL
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB  ...    5000             1
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR  ...  108590             1
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS  ...    5000             0
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION  ...    6692             1
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT  ...  142590             1

[5 rows x 12 columns]

In [ ]:
# look at value counts of STATUS column to see if any opportunity for optimization
application_df['STATUS'].value_counts()

1    34294
0        5
Name: STATUS, dtype: int64

In [ ]:
# look at value counts of SPECIAL_CONSIDERATIONS column to see if any opportunity for optimization
application_df['SPECIAL_CONSIDERATIONS'].value_counts()

N    34272
Y       27
Name: SPECIAL_CONSIDERATIONS, dtype: int64

We will be dropping both the STATUS and SPECIAL_CONSIDERATIONS columns due to the majority of results heavily favoring one outcome and not providing any really significant input to our model.

In [ ]:
# Drop the non-beneficial EIN column
application_df = application_df.drop(['EIN'], 1)

# Drop the STATUS and SPECIAL_CONSIDERATIONS columns as well
application_df = application_df.drop(['STATUS', 'SPECIAL_CONSIDERATIONS'], 1)
application_df.head()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only
  
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:5: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only
  """


NAME  ... IS_SUCCESSFUL
0              BLUE KNIGHTS MOTORCYCLE CLUB  ...             1
1    AMERICAN CHESAPEAKE CLUB CHARITABLE TR  ...             1
2        ST CLOUD PROFESSIONAL FIREFIGHTERS  ...             0
3            SOUTHSIDE ATHLETIC ASSOCIATION  ...             1
4  GENETIC RESEARCH INSTITUTE OF THE DESERT  ...             1

[5 rows x 9 columns]

In [ ]:
# display the number of unique values
application_df.nunique()

NAME                19568
APPLICATION_TYPE       17
AFFILIATION             6
CLASSIFICATION         71
USE_CASE                5
ORGANIZATION            4
INCOME_AMT              9
ASK_AMT              8747
IS_SUCCESSFUL           2
dtype: int64

In [ ]:
# Look at NAME value counts for binning
namecounts = application_df['NAME'].value_counts()
namecounts

PARENT BOOSTER USA INC                                                  1260
TOPS CLUB INC                                                            765
UNITED STATES BOWLING CONGRESS INC                                       700
WASHINGTON STATE UNIVERSITY                                              492
AMATEUR ATHLETIC UNION OF THE UNITED STATES INC                          408
                                                                        ... 
ST LOUIS SLAM WOMENS FOOTBALL                                              1
AIESEC ALUMNI IBEROAMERICA CORP                                            1
WEALLBLEEDRED ORG INC                                                      1
AMERICAN SOCIETY FOR STANDARDS IN MEDIUMSHIP & PSYCHICAL INVESTIGATI       1
WATERHOUSE CHARITABLE TR                                                   1
Name: NAME, Length: 19568, dtype: int64

In [ ]:
#  How many name counts are greater than 5?
namecounts[namecounts > 5]

PARENT BOOSTER USA INC                             1260
TOPS CLUB INC                                       765
UNITED STATES BOWLING CONGRESS INC                  700
WASHINGTON STATE UNIVERSITY                         492
AMATEUR ATHLETIC UNION OF THE UNITED STATES INC     408
                                                   ... 
OLD OAK CLIFF CONSERVATION LEAGUE INC                 6
AMERICAN NEPHROLOGY NURSES ASSOCIATION                6
HUMBLE ISD EDUCATIONAL SUPPORT GROUPS INC             6
PROFESSIONAL LOADMASTER ASSOCIATION                   6
CBMC INC                                              6
Name: NAME, Length: 354, dtype: int64

In [ ]:
# Determine which values to replace if counts are less than or equal to 5.
number_to_replace = list(namecounts[namecounts <= 5].index)

# Replace in dataframe
for name in number_to_replace:
    application_df['NAME'] = application_df['NAME'].replace(name,"Other")
    
# Check to make sure binning was successful
application_df['NAME'].value_counts()

Other                                           20043
PARENT BOOSTER USA INC                           1260
TOPS CLUB INC                                     765
UNITED STATES BOWLING CONGRESS INC                700
WASHINGTON STATE UNIVERSITY                       492
                                                ...  
HABITAT FOR HUMANITY INTERNATIONAL                  6
DAMAGE PREVENTION COUNCIL OF TEXAS                  6
FLEET RESERVE ASSOCIATION                           6
HUGH OBRIAN YOUTH LEADERSHIP                        6
INTERNATIONAL CONGRESS OF CHURCHES MINISTERS        6
Name: NAME, Length: 355, dtype: int64

In [ ]:
# Look at APPLICATION_TYPE value counts for binning
Appcounts = application_df['APPLICATION_TYPE'].value_counts()
Appcounts

T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64

In [ ]:
# Determine which values to replace if application counts are less than 500
apps_to_replace = list(Appcounts[Appcounts < 500].index)

# Replace in dataframe
for apps in apps_to_replace:
    application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(apps,"Other")

# Check to make sure binning was successful
application_df['APPLICATION_TYPE'].value_counts()

T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
Other      276
Name: APPLICATION_TYPE, dtype: int64

In [ ]:
# Look at CLASSIFICATION value counts for binning
Classcounts = application_df['CLASSIFICATION'].value_counts()
Classcounts

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: CLASSIFICATION, Length: 71, dtype: int64

In [ ]:
# Determine which values to replace if counts are less than 1000
class_to_replace = list(Classcounts[Classcounts < 1000].index)

# Replace in dataframe
for cls in class_to_replace:
    application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls,"Other")

# Check to make sure binning was successful
application_df['CLASSIFICATION'].value_counts()

C1000    17326
C2000     6074
C1200     4837
Other     2261
C3000     1918
C2100     1883
Name: CLASSIFICATION, dtype: int64

In [ ]:
# Generate our categorical variable lists
categories = application_df.dtypes[application_df.dtypes == 'object'].index.tolist()
categories

['NAME',
 'APPLICATION_TYPE',
 'AFFILIATION',
 'CLASSIFICATION',
 'USE_CASE',
 'ORGANIZATION',
 'INCOME_AMT']

In [ ]:
# perform one-hot encoding on the data
dummiesDF = pd.get_dummies(application_df)
dummiesDF.head()

ASK_AMT  IS_SUCCESSFUL  ...  INCOME_AMT_50M+  INCOME_AMT_5M-10M
0     5000              1  ...                0                  0
1   108590              1  ...                0                  0
2     5000              0  ...                0                  0
3     6692              1  ...                0                  0
4   142590              1  ...                0                  0

[5 rows x 396 columns]

In [ ]:
# Split our preprocessed data into our features and target arrays
X = dummiesDF.drop(['IS_SUCCESSFUL'], axis='columns').values
y = dummiesDF['IS_SUCCESSFUL'].values

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test, = train_test_split(X, y)

In [ ]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

In [ ]:
# Define the model - deep neural net
number_input_features = len(X_train[0])
hidden_nodes_layer1 =  100
hidden_nodes_layer2 = 30
hidden_nodes_layer3 = 10

nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer1, input_dim=number_input_features, activation="relu"))

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer2, activation="sigmoid"))

# Third hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer3, activation="sigmoid"))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 100)               39600     
                                                                 
 dense_1 (Dense)             (None, 30)                3030      
                                                                 
 dense_2 (Dense)             (None, 10)                310       
                                                                 
 dense_3 (Dense)             (None, 1)                 11        
                                                                 
Total params: 42,951
Trainable params: 42,951
Non-trainable params: 0
_________________________________________________________________


In [ ]:
# Compile the model
nn.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

In [ ]:
# Train the model
fitmodel = nn.fit(X_train_scaled, y_train, epochs=100)

Epoch 1/100
804/804 [==============================] - 3s 3ms/step - loss: 0.5050 - accuracy: 0.7685
Epoch 2/100
804/804 [==============================] - 2s 3ms/step - loss: 0.4486 - accuracy: 0.7895
Epoch 3/100
804/804 [==============================] - 2s 3ms/step - loss: 0.4387 - accuracy: 0.7933
Epoch 4/100
804/804 [==============================] - 2s 3ms/step - loss: 0.4344 - accuracy: 0.7941
Epoch 5/100
804/804 [==============================] - 2s 3ms/step - loss: 0.4313 - accuracy: 0.7951
Epoch 6/100
804/804 [==============================] - 2s 3ms/step - loss: 0.4290 - accuracy: 0.7949
Epoch 7/100
804/804 [==============================] - 2s 3ms/step - loss: 0.4273 - accuracy: 0.7966
Epoch 8/100
804/804 [==============================] - 2s 3ms/step - loss: 0.4260 - accuracy: 0.7972
Epoch 9/100
804/804 [==============================] - 2s 3ms/step - loss: 0.4249 - accuracy: 0.7967
Epoch 10/100
804/804 [==============================] - 2s 3ms/step - loss: 0.4240 - accura

In [ ]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 1s - loss: 0.4480 - accuracy: 0.7979 - 532ms/epoch - 2ms/step
Loss: 0.447968453168869, Accuracy: 0.7979008555412292


In [ ]:
# Export our model to HDF5 file
nn.save("AlphabetSoupCharity_Optimization.h5")

USE A RANDOM FOREST CLASSIFIER ON THE DATASET

In [ ]:
from sklearn.metrics import accuracy_score
from sklearn.ensemble import RandomForestClassifier

In [ ]:
# Create a random forest classifier.
rf = RandomForestClassifier(n_estimators=128, random_state=78)

# Fitting the model
rf = rf.fit(X_train_scaled, y_train)

# Evaluate the model
ypred = rf.predict(X_test_scaled)
print(f'Random Forest Model Accuracy: {accuracy_score(y_test, ypred) * 100:.2f}%')

Random Forest Model Accuracy: 77.90%


# Step 4
1.	**Overview**:
> The purpose of this analysis was to help the non-profit foundation Alphabet Soup to create an algorithm to predict whether or not applicants for funding will be successful.



2.	**Results**:
*   *Data Preprocessing*

  o	What variable(s) are considered the target(s) for your model?
> The IS_SUCCESSFUL column is binary and will be the target variable or output for our model.

  o	What variable(s) are considered to be the features for your model?
> The remaining columns after we drop EIN and NAME will be the features or input for our model.
  *   AFFILIATION
  *   CLASSIFICATION
  *   USE_CASE
  *   ORGANIZATION
  *   STATUS
  *   INCOME_AMT
  *   SPECIAL_CONSIDERATIONS
  *   ASK_AMT

  o	What variable(s) are neither targets nor features, and should be removed from the input data?
> The EIN and NAME columns can be removed from the inputs in order to optimize the results.

*   *Compiling, Training, and Evaluating the Model*

  o	How many neurons, layers, and activation functions did you select for your neural network model, and why?
> The number of neurons selected was 80/30 for the first and second layers in order to account for the number of input features after we used one-hot encoding.  The number of layers was three to account for both hidden and output layers needed to qualify as a deep learning model.  And the activation functions we used were relu and sigmoid.

  o	Were you able to achieve the target model performance?
>Yes, after optimizing the model, an accuracy score of 79.8% was achieved.

  o	What steps did you take to try and increase model performance?
> In order to optimize the model, both the STATUS and SPECIAL_CONSIDERATIONS columns were dropped due to the majority of results heavily favoring one outcome and not providing any really significant input to our model.  Also, the NAME, APPLICATION_TYPE, and CLASSIFICATION features were binned in order to reduce rare occurrences. Finally, I added another hidden layer with more neurons, and activation functions to optimize the model even further.


3.	**Summary**:
> While the first attempt did not yield an accuracy score of 75% or higher, our second attempt did by yielding an accuracy score of 79.8% after taking several steps to optimize the accuracy of the neural network model.  It is worth noting, and possibly recommended, that other models could yield sufficient results as well.  By using the Random Forest Classification model on our dataset, we immediately achieved an accuracy score of 77.9%.  This took much less time to perform while still providing acceptable results, which could be valuable in real world applications and is worth recommending.



